### PR TITLE
echo reference signal: Interface and skeleton implementation.

### PR DESCRIPTION
**Pull request by akloniex**
Echo reference signal is generated by virtual Echo Ref DAI, accompanied
with virtual buffer copier DMA. DMA will be attached to buffer component
from which echo reference signal will be collected.

As for topology, Echo reference DAI should serve as backend of timer-driven
capture pipeline, to properly report back zero-signal, while source pipeline,
to which it is attached, is paused, stopped or destroyed.

Signed-off-by: ArturX Kloniecki <arturx.kloniecki@linux.intel.com>

### CODE REVIEW

File: src/include/sof/lib/dma.h [hunk 1]
```diff
@@ -58,6 +59,7 @@ struct comp_buffer;
 #define DMA_DEV_ALH		BIT(5) /**< connectable to ALH link */
 #define DMA_DEV_SAI		BIT(6) /**< connectable to SAI fifo */
 #define DMA_DEV_ESAI		BIT(7) /**< connectable to ESAI fifo */
+#define DMA_DEV_BUFFER		BIT(8) /**< connectable to buffer component */
```
Comment (dbaluta) [inline src/include/sof/lib/dma.h]: This looks misaligned. Maybe, github interface is at fault. But please check on your side.

File: src/platform/apollolake/include/platform/lib/dma.h [hunk 2]
```diff
@@ -26,6 +26,7 @@
 #define DMA_HOST_OUT_DMAC	5
 #define DMA_LINK_IN_DMAC	6
 #define DMA_LINK_OUT_DMAC	7
+#define DMA_BUF_COPIER		8
```
Comment (dbaluta) [inline src/platform/apollolake/include/platform/lib/dma.h]: Same here, this looks misaligned.

File: src/drivers/generic/CMakeLists.txt [hunk 3]
```diff
@@ -3,3 +3,7 @@
 if (CONFIG_DUMMY_DMA)
 	add_local_sources(sof dummy-dma.c)
 endif()
```
Comment (dbaluta) [inline src/drivers/generic/CMakeLists.txt]: Please split this into at least 3 commits:
1) Adding the buffer copier interface
2) Adding the echo reference interface
3) The rest.

File: src/drivers/generic/buf-copier-dma.c [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Artur Kloniecki <arturx.kloniecki@linux.intel.com>
+//
+
+ /**
+  * \file drivers/generic/buf-copier-dma.c
+  * \brief Driver for Virtual DMA with buffer copying capability.
+  * \author Artur Kloniecki <arturx.kloniecki@linux.intel.com>
+  */
```
Comment (lgirdwood) [inline src/drivers/generic/buf-copier-dma.c]: This needs more context on what, why etc.

File: src/include/sof/lib/dma.h (hunk 1 above)
Comment (akloniex) [reply src/include/sof/lib/dma.h]: @dbaluta I've double-checked, it's properly aligned in the code. It's just a GH quirk.

File: src/platform/apollolake/include/platform/lib/dma.h (hunk 2 above)
Comment (akloniex) [reply src/platform/apollolake/include/platform/lib/dma.h]: @dbaluta Same as the above, it's just GH display issue.

**Comment by akloniex** [comment]
@lgirdwood I suppose that the topology should be sketched as ASCII art? I'm not very familiar with actual topology .m4 files.

**Comment by lgirdwood** [comment]
> @lgirdwood I suppose that the topology should be sketched as ASCII art? I'm not very familiar with actual topology .m4 files.

ASCII art is always good. :)

**Comment by dbaluta** [comment]
@akloniex for ASCII art you can use online editor from here: http://asciiflow.com/

**Comment by akloniex** [comment]
```
Generic Playback pipeline
+----------------------------------------------------+
|                                                    |
| +------+    +-----------+    +--------+    +-----+ |
| |      |    |           |    |        |    |     | |
| | Host +--->+ Prcessing +--->+ Buffer +--->+ DAI | |
| |      |    |           |    |        |    |     | |
| +------+    +-----------+    +--------+    +-----+ |
|                                  ||                |
+----------------------------------------------------+
                                   ||
                                   ||
                                   || Buf Copier DMA
                                   ||
                                   ||
                 +--------------------------------------------+
                 |                 ||                         |
                 | +------+    +---vv---+    +--------------+ |
                 | |      |    |        |    |              | |
                 | | Host +<---+ Buffer +<---+ Echo Ref DAI | |
                 | |      |    |        |    |              | |
                 | +------+    +--------+    +--------------+ |
                 |                                            |
                 +--------------------------------------------+
                 Echo Reference Capture pipeline
```
Exact same scheme was added to one of commit messages.

File: src/drivers/generic/CMakeLists.txt (hunk 3 above)
Comment (akloniex) [reply src/drivers/generic/CMakeLists.txt]: Split into series of commits.

**Comment by paulstelian97** [comment]
Minor comment: For the buffer copier vdma, can't part of the Dummy DMA driver be reused? It is after all capable of doing memory-to-memory copies. (of course the PR title says we just have a skeleton, but since code reuse wouldn't be that bad of an idea I wanted to point this out)

**Comment by dbaluta** [comment]
@paulstelian97 actually I think that's a good idea. We can then rename dummy-dma.c to vdma.c. I don't know exactly the inner details but it looks the same scenario -> memory to memory copy.

**Comment by lgirdwood** [comment]
@akloniex IIUC, the "copier" is taking stereo from the playback stream and copying into an already existing buffer ? I see a different topology that means we dont have to make buffers into mixing/muxing components.i.e.
```
Host ---> Proc----> Selector0 --> Buf0 ---- > DAI Playback
                      |
                      v
                     Buf1
                      |
                      v
Host <---------------Selector1 <----- Buf2 <----- Echo Ref DAI
```
This means "selector0" takes and duplicates its input data to 2 or more output buffers. Whilst selector1 does the opposite, takes channels from multiple inputs and puts them in one output buffer (using channel map). This may need a small update to the selector code but it's mostly there.

**Comment by akloniex** [comment]
@lgirdwood IIUC, the idea behind using vdai and vdma was to avoid intertwined pipelines, component-wise, to avoid xrun situations when playback pipeline is stopped. Also the whole issue with params and triggers propagation is cut short, with 2 topologically separate pipelines.
In the past we have delivered topology similar to the one that You suggest, with Demux component routing data back on capture pipeline, although it proved to cause some issues on kernel side due to the capture pipeline being 'dailess'. Dummy dai could be partial solution to the problem, if we replaced Selector0 in your scheme with Demux, and Selector1 with Mux, but that still wouldn't solve xruns, in case of one of the pipelines stops or pauses.
Keeping the pipelines topologically separate solves that, at cost of complexity under the hood.

**Comment by lgirdwood** [comment]
@akloniex using the mux/demux/selector here below does mean 2 pipelines, but no hostless DAI or complexity under the hood if we update the (mux/demux/selector) i.e.

1) buf 1 is the pipe0 -> pipe1 boundary, it can exist in pipe0 but it does not propagate params (from pipe0 into pipe1)

2) Selector0 should not complain (and set XRUN) if Buf1 overflows, it should operate as normal as long as 1 output is operating normally.

3) Selector1 should not emist XRUNs if Buf1 underflows, it should operate normally as long as 1 input is operating normally.

So some logic need to be added to the (mux/demux/selector) for usage in the operating mode. This keeps the complexity out of the buffers, pipeline and other logic.

```
Host ---> Proc----> Selector0 --> Buf0 ---- > DAI Playback
                      |
                      v
                     Buf1
                      |
                      v
Host <---------------Selector1 <----- Buf2 <----- Echo Ref DAI
```

**Comment by akloniex** [comment]
@lgirdwood I believe there are few reasons why we should go with the proposed solution:
- we have released it as of cml-006-drop-stable, which went stale a long time ago in PR https://github.com/thesofproject/sof/pull/1652
- doesn't require any functional or interface changes of components already in the codebase
- doesn't require turning xruns off for said components
- most of the implementation is already done I just need few final touch-ups before sending another PR with the implementation

**Comment by lgirdwood** [comment]
@akloniex I dont want to overload the buffer APIs ( where we have to track multiple readers/writers for one use case) and add unnecessary logic that is specific to a single use case. Switching XRUNs off per component is actually quite useful for pipelines that are allowed to stall. This feature can also be achieved using existing components without adding new ones.

In general, it's always better to reuse existing code (mux / demux / selector) and extend the current component API (switch off XRUN handling) than inventing something new that is specific for a single use case. 

I also do not see where we have to extend the kernel/userspace here with the exception of adding a component tuple to disable XRUN handling.

**Comment by akloniex** [comment]
@lgirdwood So essentionally, IIUC, you're suggesting putting a tone generator, with zero amplitude in place of DAI and providing echo reference signal generation by the means of topology, rather than custom solution, correct? Because I don't think that creating echo ref DAI just for the sake of generation of time signature is really necessary here, if we follow your suggestion one step further.

**Comment by lgirdwood** [comment]
> @lgirdwood So essentionally, IIUC, you're suggesting putting a tone generator, with zero amplitude in place of DAI and providing echo reference signal generation by the means of topology, rather than custom solution, correct? Because I don't think that creating echo ref DAI just for the sake of generation of time signature is really necessary here, if we follow your suggestion one step further.

Sorry, not following you here ? I'm suggesting we support loopback using regular components (that have a topology flag to ignore xruns).

**Comment by akloniex** [comment]
@lgirdwood That part seems clear to me. What I was getting at, is that having Echo Reference DAI seems obsolete in such solution, and I would guess that we could replace it with Tone Generator, as scheduling component for the echo-capture pipeline. Or am I missing something here?

**Comment by lgirdwood** [comment]
> @lgirdwood That part seems clear to me. What I was getting at, is that having Echo Reference DAI seems obsolete in such solution, and I would guess that we could replace it with Tone Generator, as scheduling component for the echo-capture pipeline. Or am I missing something here?

Ok, so why is ER DAI obsolete ? Isn't this just a regular DAI with a Mic/DMIC ? Do you want to use the Tone comp for testing purposes only ?

**Comment by akloniex** [comment]
@lgirdwood Wasn't echo reference signal supposed to provide reference of what is being played by the playback pipeline, to allow for any echo cancellation to be done by software? The DAI isn't supposed to generate/gather any signal, just serve as backend for the pipeline.
In my solution, DAI served as scheduling component, and as proxy to DMA, that was doing all the lifting.
In your proposed solution it doesn't play any role on it's own.

**Comment by lgirdwood** [comment]
@akloniex sorry, now really not following you. Both our pipelines move the data in the same directions. We both have 2 pipelines in our proposals. I dont think we have any difference here on data flow or pipeline direction ?
I think our difference is getting the data from payback pipeline to ER pipeline ? You want to do buffer -> buffer copy via virtual DMA (new logical pipeline flow) whereas I'm suggesting we reuse current comp -> buffer -> comp pipeline flow.

**Comment by akloniex** [comment]
@lgirdwood As we've discussed during our call, I've split the previous diagram into 2 separate ones, to better visualize the 2 types of information contained.

Topology - 2 separate pipelines:
```
Playback pipeline
+------------------------------------------+
|                                          |
| +----+   +----------+   +------+   +---+ |
| |Host+-->+Processing+-->+Buffer+-->+DAI| |
| +----+   +----------+   +------+   +---+ |
|                                          |
+------------------------------------------+

+------------------------------------+
|                                    |
| +----+   +------+   +------------+ |
| |Host+<--+Buffer+<--+Echo Ref DAI| |
| +----+   +------+   +------------+ |
|                                    |
+------------------------------------+
Echo Reference pipeline
```

Data Flow - data duplicated from buffer in Playback pipeline into Echo Ref by Buf Copier DMA:
```
Playback pipeline
+------------------------------------------+
|                                          |
| +----+   +----------+   +------+   +---+ |
| |Host+-->+Processing+-->+Buffer+-->+DAI| |
| +----+   +----------+   +------+   +---+ |
|                            |             |
+------------------------------------------+
                             |
                             | Buf Copier DMA
                             |
               +-------------------+
               |             |     |
               | +----+   +--v---+ |
               | |Host+<--+Buffer| |
               | +----+   +------+ |
               |                   |
               +-------------------+
               Echo Reference pipeline
```

Hope that this fully clarifies the situation, and I'm hoping for getting green light from you now.

**Comment by lgirdwood** [comment]
@akloniex thanks, seeing your flow better now. IIRC, the echo ref DAI is only used as a time source to schedule the echo reference pipeline ? We would not need the Echo ref DAI or the Copier DMA with the following flow 
```
Host --> buf0 --> Processing --> buf1 --> demux--+--> buf2 --> DAI
                                                 |
Host <----------------------------------- buf3 --+
```
_Single pipeline where buf3 allows underrunn and overrun._

This pipeline above uses existing components. It support reuse, and the copying is all synchronised via the DAI (i.e. for host playback and capture).

**Comment by akloniex** [comment]
@ligrdwood Correct, but if I remember correctly, we've already delivered such solution way back, as of cml-004 to be exact. @juimonen created such topology, fighting some issues on kernel side with dailess pipeline, but finally solved that with some virtual widgets, IIRC. Although, back then we didn't have the xrun prevention which You mentioned.
As @xiulipan noticed, that solution also has the disadvantage of being statically defined, while with completely separate pipeline we could create such pipeline dynamically, connected to any pipe we wish, which makes the new solution more flexible.

**Comment by lgirdwood** [comment]
> @ligrdwood Correct, but if I remember correctly, we've already delivered such solution way back, as of cml-004 to be exact.

Sorry, just because it was implemented in a certain way for production release in the past, in no way will guarantee that it will be accepted in master. This is true for most if open source projects. 

 @juimonen created such topology, fighting some issues on kernel side with dailess pipeline, but finally solved that with some virtual widgets, IIRC. Although, back then we didn't have the xrun prevention which You mentioned.
> As @xiulipan noticed, that solution also has the disadvantage of being statically defined, while with completely separate pipeline we could create such pipeline dynamically, connected to any pipe we wish, which makes the new solution more flexible.

This is a separate discussion an unrelated to this use case. Pipeline definitions can today be dynamically updated using the current APIs.

As discussed your proposal has big flaws.

1) Uses a different DAI as a time source, adding complexity and meaning the DAI can't be used by other pipelines for audio. This can also get far more complex if your only available DAIs are on a different time domain.

2) The "virtual DMA" is not synchronised between both of your pipelines. It's non deterministic as the host can trigger pipeline1 & 2 independently, which will follow on that the copy between 1 & 2 will retain host derived timing differences.

3) The virtual DMA is a second reader on a buffer. This is generally OK if both readers are exactly frame synchronised, but 1 & 2 above we are not frame synchronised.

**Comment by lbetlej** [comment]
@lgirdwood Could you describe your idea to simplify complexity on FW side and move complexity to upper layers of audio stack ?

**Comment by lgirdwood** [comment]
> @lgirdwood Could you describe your idea to simplify complexity on FW side and move complexity to upper layers of audio stack ?

We should always move any complexity to the correct place in the stack. 

Putting this complexity in the FW means we the three reasons above (+ extra PM cost of running another DAI/DMA). However, putting this logic in sound sever and UCM means re-ordering the trigger() order of playback and capture PCM device. The latter option is much easier.

Feature created https://github.com/thesofproject/sof/issues/2564

**Comment by plbossart** [comment]
Way too optimistic @lgirdwood... The state of userspace is a mess. we can't get jack detection right so what are the odds we could model dependencies between streaming PCMs?
Yes we should move the complexity to the highest level we can control and influence. We can't depend on something as flaky as UCM for PCM synchronization.

**Comment by lgirdwood** [comment]
@plbossart way too pessimistic :-) well lets improve userspace and incentivize others to contribute in this space too (as it's far easier for the community to get involved at this level). The first steps are really low hanging fruit here (which can then let folks like Jaroslav build on top) 

1) The UCM tuple for PCM dependence as @cujomalainey and I have mentioned.
2) Echo ref (and similar PCM) to return -ENODATA on trigger if their dependant PCM is inactive (FW can return this value as part of trigger IPC, driver just passes up the stack).
3) Update aplay/arecord to print appropriate and helpful error if PCM returns ENODATA.
4) CRAS to use UCM tuple
5) Pulseaudio to log appropriate message if PCM trigger returns ENODATA. It could also lookup UCM tuple and use the UCM values in the log message.

All the above are easy todo and allow others to build on.

**Comment by plbossart** [comment]
> @plbossart way too pessimistic :-) well lets improve userspace and incentivize others to contribute in this space too (as it's far easier for the community to get involved at this level). The first steps are really low hanging fruit here (which can then let folks like Jaroslav build on top)
> 
> 1. The UCM tuple for PCM dependence as @cujomalainey and I have mentioned.
> 2. Echo ref (and similar PCM) to return -ENODATA on trigger if their dependant PCM is inactive (FW can return this value as part of trigger IPC, driver just passes up the stack).

wrong layer, the kernel has no concept of UCM. This has to be a topology-based definition.

> 3. Update aplay/arecord to print appropriate and helpful error if PCM returns ENODATA.
> 4. CRAS to use UCM tuple
> 5. Pulseaudio to log appropriate message if PCM trigger returns ENODATA. It could also lookup UCM tuple and use the UCM values in the log message.
> 
> All the above are easy todo and allow others to build on.

**Comment by lgirdwood** [comment]
> > 2\. Echo ref (and similar PCM) to return -ENODATA on trigger if their dependant PCM is inactive (FW can return this value as part of trigger IPC, driver just passes up the stack).
> 
> wrong layer, the kernel has no concept of UCM. This has to be a topology-based definition.

The kernel wont need to do anything here, all it has to do is pass up the -ENODATA from FW to userspace from trigger(START). 

FW will know to return -ENODATA if it walks the graph and finds no DAI endpoints and inactive PCM endpoints.

**Pull request closed without merging**